[Bug][core] make multi_layer_kv_transfer() deal with flashinfer and triton

ziruiliu opened the pull request:
**Why**
As described in #2291 ,  `multi_layer_kv_transfer()` in mem_kernels.cu assumed a single vLLM layout and produced incorrect offsets when vLLM uses [num_blocks, 2, block_size, ...]  causing data copy errors with flashinfer or triton
Single-layer transfers `single_layer_kv_transfer()` already had a vllm_two_major flag,  so the multi-layer path and GPU connectors needed the same awareness to handle both layouts correctly.

**What**
- Added `vllm_two_major` and `block_size` parameters to the `multi_layer_kv_transfer` declaration in `csrc/mem_kernels.cuh` and threaded them through the kernel launches in `csrc/mem_kernels.cu`.
- Made `page_buffer_offset` layout-aware by updating its signature and logic to compute offsets for both two-major and block-major layouts, and passed `block_size`/`vllm_two_major` into the `load_and_reshape_multi_layer_kernel`.
- Updated GPU connectors (`VLLMPagedMemGPUConnectorV2`, `VLLMPagedMemGPUConnectorV3`) to detect the vLLM KV cache layout and compute `page_buffer_size` and `block_size`, and pass `self.vllm_two_major`/`self.block_size` into `lmc_ops.multi_layer_kv_transfer` calls.
- Updated `GPUCacheContext` in `lmcache/v1/multiprocess/server.py` to detect layout (`vllm_two_major_`) and pass `vllm_two_major`/`block_size` through the places that call `lmc_ops.multi_layer_kv_transfer`.
- Updated `test_mem_kernels.py` accordingly

ziruiliu committed:
Enable multi_layer_kv_transfer() deal with kv
cache shape starts with num_blocks for flashinfer
and triton

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ziruiliu committed:
update comment

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ApostaC commented:
Hey @ziruiliu , appreciate the contribution!

I do have a concern about the code. As the arguments become more and more, it becomes very hard to track the meaning of each argument (e.g., it's not clear what `vllm_two_major` means if we only read the variable name), and it creates a lot of confusion in the Python layer when seeing a list of `True/False` arguments.

Therefore, I would like to see another non-functionality refactoring to export enum definition from c-level to python-level via pybind.

I'm open to offline discussions to move this forward. Please let me know!

ziruiliu commented:
> Hey @ziruiliu , appreciate the contribution!
> 
> I do have a concern about the code. As the arguments become more and more, it becomes very hard to track the meaning of each argument (e.g., it's not clear what `vllm_two_major` means if we only read the variable name), and it creates a lot of confusion in the Python layer when seeing a list of `True/False` arguments.
> 
> Therefore, I would like to see another non-functionality refactoring to export enum definition from c-level to python-level via pybind.
> 
> I'm open to offline discussions to move this forward. Please let me know!

Totally agree. My PR is simply following the style of single layer functions those using vllm_two_major. Do you want me to add some enum definition and also apply them to existing single layer transfer functions?

ziruiliu commented:
> Hey @ziruiliu , appreciate the contribution!
> 
> I do have a concern about the code. As the arguments become more and more, it becomes very hard to track the meaning of each argument (e.g., it's not clear what `vllm_two_major` means if we only read the variable name), and it creates a lot of confusion in the Python layer when seeing a list of `True/False` arguments.
> 
> Therefore, I would like to see another non-functionality refactoring to export enum definition from c-level to python-level via pybind.
> 
> I'm open to offline discussions to move this forward. Please let me know!

Let's pass the flags and options this way:
1. Use enum Direction::D2H and Direction::H2D to replace current True/False direction.
2. Use KVLayout::MLA, KVLayout::BlockFirst (shape starts with num_blocks) and KVLayout::KVFirst (shape starts with 2) to replace current True/False flags.

@ApostaC I will soon prepare an update. If you have any other idea please let me know. Thanks

ziruiliu committed:
revert mem_kernels.cu to latest base

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ziruiliu committed:
use enums instead of bool values

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ziruiliu committed:
update server.py's  multi_layer_kv_transfer usage accordingly

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ziruiliu committed:
update server.py

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ziruiliu committed:
Merge branch 'dev' into fix/attn_shape

ziruiliu committed:
Merge branch 'dev' into fix/attn_shape

ziruiliu commented:
updated to enum, but still working on unit test failures

ziruiliu committed:
update for unit test

Signed-off-by: Zirui Liu <ziliu@ddn.com>

ziruiliu commented:
Hi @ApostaC , I have updated this fix, introduced two enums:
1. TransferDirection::D2H and H2D to replace True/False for read/write
2. KVLayout::MLA, KVFirst and BlockFirst to represent different kv cache shapes, to replace bool value vllm_two_major

Please take a look at it, thanks.

ApostaC commented:
@ziruiliu Hey Zirui, thanks for the update! I will take a look at it soon!

ApostaC commented:
Hey @ziruiliu , just wanted to let you know that I have picked a few lines of your changes in this PR into #2413

sammshen commented:
Hey @ziruiliu , just wanted to let you know that I will pick a few lines of your changes in this PR into https://github.com/LMCache/LMCache/pull/2518

ziruiliu commented:
> Hey @ziruiliu , just wanted to let you know that I will pick a few lines of your changes in this PR into #2518

Got it. I understand that the community has its own priorities to refactor current code. Please feel free to use code from this PR

ziruiliu commented:
I believe the issue this PR trying to address is already fixed in latest LMCache
